### ISSUE TITLE
Can't use dockers on two machines

### ISSUE THREAD

**Issue by ksy3395** [horovod/horovod#397]
Hello, 

I got issues that I can't test horovod with docker on two machines. They are frozen with these commands:

* Master:
(tensorflow) root@master:/examples# mpirun -np 2 -H 142.93.31.180:1,142.93.31.194:1 -bind-to none -map-by slot -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -mca plm_rsh_args "-p 12345" python tensorflow_mnist.py
Warning: Permanently added '[142.93.31.194]:12345' (ECDSA) to the list of known hosts.
^C(tensorflow) root@master:/examples# mpirun -np 2 -H 142.93.31.180:1,142.93.31.194:1 -bind-to none -map-by slot -x NCCDEBUG=INFO -x LD_LIBRARY_PATH -mca plm_rsh_args "-p 12345" -vvvv 142.93.31.194 python tensorflow_mnist.py

(No response)

* Slave:
root@slave:~# docker run -it --network=host -v /mnt/share/ssh:/root/.ssh horovod:latest bash -c "/usr/sbin/sshd -p 125; sleep infinity"

(No response)

Could you let me know what I could try this case?

Thank you,
Soyoung

**Comment by alsrgv** [comment]
@ksy3395, can you check whether your machines have any non-routed network interfaces and if so, exclude those? See https://github.com/uber/horovod/blob/master/docs/running.md#hangs-due-to-non-routed-network-interfaces.

**Comment by ksy3395** [comment]
Hello Alex, 

Thank you for your time and help.

When I checked it with ifconfig, the result was:

* root@master:~# ifconfig
docker0   Link encap:Ethernet  HWaddr 02:42:d3:e1:df:c4  
          inet addr:172.17.0.1  Bcast:172.17.255.255  Mask:255.255.0.0
          inet6 addr: fe80::42:d3ff:fee1:dfc4/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:51060 errors:0 dropped:0 overruns:0 frame:0
          TX packets:84805 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:2706676 (2.7 MB)  TX bytes:1089089240 (1.0 GB)

eth0      Link encap:Ethernet  HWaddr 06:ff:56:ee:be:61  
          inet addr:142.93.31.180  Bcast:142.93.31.255  Mask:255.255.240.0
          inet6 addr: fe80::4ff:56ff:feee:be61/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:305342 errors:0 dropped:0 overruns:0 frame:0
          TX packets:256461 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:2562661010 (2.5 GB)  TX bytes:34039206 (34.0 MB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:185 errors:0 dropped:0 overruns:0 frame:0
          TX packets:185 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 

* root@slave:~# ifconfig
docker0   Link encap:Ethernet  HWaddr 02:42:0d:98:08:6d  
          inet addr:172.17.0.1  Bcast:172.17.255.255  Mask:255.255.0.0
          inet6 addr: fe80::42:dff:fe98:86d/64 Scope:Link
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:41162 errors:0 dropped:0 overruns:0 frame:0
          TX packets:65425 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:2240737 (2.2 MB)  TX bytes:1086061661 (1.0 GB)

eth0      Link encap:Ethernet  HWaddr b2:e2:f0:52:7b:1d  
          inet addr:142.93.31.194  Bcast:142.93.31.255  Mask:255.255.240.0
          inet6 addr: fe80::b0e2:f0ff:fe52:7b1d/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:324311 errors:0 dropped:0 overruns:0 frame:0
          TX packets:293790 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:2334329811 (2.3 GB)  TX bytes:30223435 (30.2 MB)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:160 errors:0 dropped:0 overruns:0 frame:0
          TX packets:160 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1 
          RX bytes:11840 (11.8 KB)  TX bytes:11840 (11.8 KB)
--------------------------------------------------------------

So I think it has non-routed network interface, but I still didn't get any responses with -mca btl_tcp_if_exclude <interface>[,<interface>] and NCCL_SOCKET_IFNAME=^<interface>.

The result was:
* root@master:/examples# mpirun -np 2 -H 142.93.31.180:1,142.93.31.194:1 -bind-to none -map-by slot -x -x NCCL_SOCKET_IFNAME=^docker0 -mca btl_tcp_if_exclude lo,docker0 NCCDEBUG=INFO -x LD_LIBRARY_PATH -mca plm_rsh_args "-p 12345" -vvvv 142.93.31.194 python tensorflow_mnist.py
[master:00012] Warning: could not find environment variable "-x"

* root@slave:~# # docker run -it --network=host -v /mnt/share/ssh:/root/.ssh horovod:latest -x NCCL_SOCKET_IFNAME=^docker0 bash -c "/usr/sbin/sshd -p 125; sleep infinity"

Thank you so much!
Soyoung

**Comment by alsrgv** [comment]
Can you fix the first command to be:
```
(inside docker)# mpirun -np 2 -H 142.93.31.180:1,142.93.31.194:1 -bind-to none -map-by slot -x NCCL_SOCKET_IFNAME=^docker0 -mca btl_tcp_if_exclude lo,docker0 -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -mca plm_rsh_args "-p 12345 -vvvv" python tensorflow_mnist.py
```

And the second command to be:
```
(on the second host)$ docker run -it --network=host -v /mnt/share/ssh:/root/.ssh horovod:latest bash -c "/usr/sbin/sshd -p 12345; sleep infinity"
```

Please diff them vs your commands to discover the differences.

**Comment by vilmara** [comment]
hi @ksy3395 I had the same issue (among other issues) and @alsrgv helped me to fix it on the issue #288. He realized that the container on the second node was running an older version of the image. So I rebooted the second server as recommended by Alex, and it worked.
